### PR TITLE
Shared Persistent Volume

### PR DESCRIPTION
**Pull request by kevtaylor**
SharedPersistentVolume=true/false in cluster.yaml
Creates a dynamic EFS volume available across all zones
Creates persistent volume definitions for kubernetes - derives the efs mount point
Creates service scripts to create the persistent volume in the cluster

### CODE REVIEW

**Comment by mumoshu** [comment]
Hi @kevtaylor, thanks for your contribution 👍 
Though this change looks nice, would you mind sharing me example use-cases of this feature for my education?

File: core/controlplane/config/templates/cluster.yaml [hunk 1]
```diff
@@ -775,6 +775,9 @@ worker:
 # See https://github.com/kubernetes-incubator/kube-aws/issues/208 for more information
 #elasticFileSystemId: fs-47a2c22e
 
+# Create shared persistent volume
+#sharedPersistentVolume: false
```
Comment (mumoshu) [inline core/controlplane/config/templates/cluster.yaml]: Just wondering/Not intended to scope-creep this PR; Are the any use-cases that lead us want to create multiple of shared persistent volumes, possibly varying volume names and storage sizes?

```yaml
sharedPersistentVolumes:
- name: shared-efs
  provider: efs
  size: 500Gi
- name: shared-ebs
  provider: ebs
  size: 100Gi
```

**Comment by kevtaylor** [comment]
@mumoshu The use case we have is regarding batch processes that we run using chronos api. We create indexes which ate produced by a batch job which need to be read or published by another job. An Efs volume is mounted across multiple instances so if we have processes running across a cluster they can all see the shared volume. The current kube-aws offering assumes an efs mount point so we create a dynamic one and set it up to be available in kubernetes.  The 500Gb limit could be configured but that would bloat the templates so we just figured that you could modify it to your own spec

**Comment by mumoshu** [comment]
Thanks @kevtaylor!
Are you ok with those volumes being deleted on a k8s cluster is destroyed?
In other words, don't you want to retain those efs file systems regardless of what the lifecycles of k8s clusters are?

**Comment by mumoshu** [comment]
@kevtaylor Btw, would you mind signing the CLA before merging this?

**Comment by mumoshu** [comment]
@kevtaylor Would you mind running `make format` and adds the changes from it to this PR?

**Comment by kevtaylor** [comment]
@mumoshu We understand the destruction of the volume if you take the whole cluster out and I would assume that if anyone didn't want that they could use the existing process of supplying their own I'd. We are in the process if getting the CLA requirements and will push the requested format shortly.

File: core/controlplane/config/templates/cloud-config-controller [hunk 2]
```diff
@@ -345,6 +345,23 @@ coreos:
...
+        Wants=kube-node-taint-and-uncordon.service
+        After=kube-node-taint-and-uncordon.service
+        [Service]
+        Type=simple
+        RemainAfterExit=true
+        RestartSec=10
+        Restart=on-failure
+        ExecStartPre=/opt/bin/set-efs-pv
+        ExecStartPre=/usr/bin/systemctl is-active kube-node-taint-and-uncordon.service
+        ExecStart=/opt/bin/load-efs-pv
```
Comment (mumoshu) [inline core/controlplane/config/templates/cloud-config-controller]: Perhaps you're missing an entry for the `/opt/bin/load-efs-pv` script under `write_files:`?

Comment (kevtaylor) [inline core/controlplane/config/templates/cloud-config-controller]: My colleague pushed the missing part up to the commit. Apologies.

Comment (mumoshu) [inline core/controlplane/config/templates/cloud-config-controller]: Thanks! Once other minor comments are addressed, I'll be ready to merge this

**Comment by mumoshu** [comment]
@kevtaylor Could you reply with `I signed it!` so that the CI verifies your organization's sign on CLA?

**Comment by mumoshu** [comment]
Hi @jollinshead! Is @kevtaylor your colleague? Then, can you confirm that you have signed CLA as an organization?

**Comment by kevtaylor** [comment]
@mumoshu Hi. I am waiting for a representative from our organisation to sign the CLA. I'll post here when I get it. Sorry for the delay but politics gets in the way of progress :-)

**Comment by mumoshu** [comment]
@kevtaylor Got it. Thanks for the reply and sorry for rushing you 👍

**Comment by kevtaylor** [comment]
I have been approved as an individual

**Comment by mumoshu** [comment]
Thanks!
I'd appreciate it if you could resolve conflicts
2017年4月13日(木) 18:15 Kevin Taylor <notifications@github.com>:

**Comment by mumoshu** [comment]
@kevtaylor LGTM. Thanks for your contribution 👍

**Pull request merged**